File: src/main/java/com/bajdcc/LALR1/grammar/tree/closure/ClosureScope.java
package com.bajdcc.LALR1.grammar.tree.closure;

import java.util.HashSet;
import java.util.Set;

/**
 * 闭包
 *
 * @author bajdcc
 */
public class ClosureScope implements IClosureScope {

	private Set<Object> ref = new HashSet<>();
	private Set<Object> decl = new HashSet<>();

	@Override
	public void addRef(Object obj) {
		ref.add(obj);
	}

	@Override
	public void addDecl(Object obj) {
		decl.add(obj);
	}

	protected Set<Object> getClosure() {
		Set<Object> closure = new HashSet<>(ref);
		closure.removeAll(decl);
		return closure.isEmpty() ? null : closure;
	}
}

File: src/main/java/com/bajdcc/LALR1/grammar/tree/closure/ClosureScope.kt
package com.bajdcc.LALR1.grammar.tree.closure

/**
 * 闭包
 *
 * @author bajdcc
 */
open class ClosureScope : IClosureScope {

    private val ref = mutableSetOf<Any>()
    private val decl = mutableSetOf<Any>()

    protected val closures: Set<Any>?
        get() {
            val closure = HashSet(ref)
            closure.removeAll(decl)
            return if (closure.isEmpty()) null else closure
        }

    override fun addRef(obj: Any) {
        ref.add(obj)
    }

    override fun addDecl(obj: Any) {
        decl.add(obj)
    }
}
